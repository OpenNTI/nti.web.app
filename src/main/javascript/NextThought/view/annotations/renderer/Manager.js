Ext.define('NextThought.view.annotations.renderer.Manager',{
	requires: [
		'NextThought.util.Line'
	],

	events: new Ext.util.Observable(),
	registry: {},
	sorter: {},
	gutter: {},
	buckets: {},
	rendererSuspended: {},

	controlLineTmpl: Ext.DomHelper.createTemplate( { cls:'controlContainer'} ).compile(),
	widgetLineTmpl: Ext.DomHelper.createTemplate( {cls:'widgetContainer'} ).compile(),
	stackTmpl: Ext.DomHelper.createTemplate( {cls:'thumb nib-stack'} ).compile(),
	addNoteToOccupiedLineTmpl: Ext.DomHelper.createTemplate( {cls:'thumb note-gutter-widget add-note {0}', 'data-qtip':'Add Note'} ).compile(),

	isDebug: false,

	/**
	 * @constructor Inner class
	 */
	Bucket: function(){
		this.values = {};
		this.length = 0;

		this.free = function(){
			this.each(function(v,k,o){ delete o[k]; },this);
			Ext.Object.each(this,function(k,o,me){ delete me[k]; });
		};

		this.put = function(o,key){
			key = typeof key === 'number'? key : guidGenerator();
			if(this.values[key]){throw 'existing value';}
			this.values[key] = o;
			this.length += 1;
			return key;
		};

		this.each = function(cb,s){
			Ext.Object.each(this.values,function(key,value,o){
				Ext.callback(cb,s,[value,key,o]);
			});
		};

		this.get = function(key){ return this.values[key]; };

		this.first = function(){
			var key;
			for (key in this.values){
				if (this.values.hasOwnProperty(key)){
					return this.get(key);
				}
			}
			return null;
		};
	},


	registerGutter: function(el, reader){
		//TODO all this junk about prefixes should go away once we aren't
		//using a singleton here...
		var p = reader.prefix;
		if(!p){ Ext.Error.raise('Prefix required'); }
		if(this.gutter[p]){
			console.warn('replacing exisiting gutter?', this.gutter[p]);
		}
		this.gutter[p] = el;

		el.controls = el.down('.controls');
		el.widgets = el.down('.widgets');

		this.render(p);//renders wait for the gutter to exist
	},


	register: function(o){
		var p = o.prefix;
		if(!this.registry[p]){
			this.registry[p] = [];
		}
		this.registry[p].push(o);
		o.requestRender();
	},


	unregister: function(o){
		var p = o.prefix, r;
		r = this.registry[p];
		if(r){
			this.registry[p] = Ext.Array.remove(r,o);
			if(this.registry[p].legend===0){
				this.sorter[p] = null;
			}
		}
	},


	getReader: function(prefix){
		var cache = this.readerPanels, c;
		if(!cache ){ cache = this.readerPanels = {}; }

		c = cache[prefix];
		if(!c){
			c = cache[prefix] = ReaderPanel.get(prefix);
			c.on('destroy',function(){ delete cache[prefix]; });
		}

		return c;
	},


	getDoc: function(prefix){
		return this.getReader(prefix).getDocumentElement();
	},


	buildSorter: function(prefix){
		//Default sort will sort by lastModified
		return function(a, b){
			var c = 0, $a = (a.record || {}).get('Last Modified'), $b = (b.record || {}).get('Last Modified');
			if($a !== $b){
				c = $a < $b? 1 : -1;
			}

			return c;
		};
	},


	clearBuckets: function(prefix){
		function clear(d){
			while(d && d.firstChild){
				d.removeChild(d.firstChild);
			}
		}

		var g = this.gutter[prefix];

		clear(g.controls.dom);
		clear(g.widgets.dom);

		if( this.buckets[prefix] ){
			this.buckets[prefix].free();
		}
		this.buckets[prefix] = new this.Bucket();
		this.getReader(prefix).noteOverlayClearRestrictedRanges();
	},


	getBucket: function(prefix, line){
//		var lineInfo,
//			originalLine = line;

		//while(!lineInfo && (originalLine - line) <= 100){
		//	lineInfo = LineUtils.findLine(line,this.getDoc(prefix));
		//	line-=5	;
		//}

//		if(!lineInfo){
//			console.error('could not resolve a line for '+prefix+' @'+line+', original line was ' + originalLine);
//			return;
//		}
		if (line < 0){
			//bad line, don't render:
			if(this.isDebug){
				console.error('Annotation cannot be rendered in gutter');
			}
			return null;
		}

		var c = this.buckets[prefix],
			lineTolerance = 40,
//			l = line.rect.top,//normalize lines
			b = c? c.get(line) : null,
			a = Math.round( line- (lineTolerance / 2)), z = Math.round( line + (lineTolerance / 2));

		if (!b && c) {
			c.each(function(value, key){
				var keyNum = parseInt(key, 10);
				if ( keyNum >=a && keyNum <= z){
					b = value;
					return false;
				}
				return true;
			});

			if (!b) {
				b = new this.Bucket();
				c.put(b,line);
			}
		}
		return b;
	},


	layoutBuckets: function(prefix){
		var addTpl = this.addNoteToOccupiedLineTmpl,
			r = this.getReader(prefix),
			g = this.gutter[prefix],
			b = this.buckets[prefix],
			cT = this.controlLineTmpl,
			wT = this.widgetLineTmpl,
			sT = this.stackTmpl,
			annotationOffsets = r.getAnnotationOffsets(),
			width = annotationOffsets.gutter + 80,
			cls = (width <= 355)? 'narrow-gutter': '',
			maxAnnotations = Math.floor(width / 34) - 3;

		g.setWidth(width);
		g.controls.setLeft(width-50);
		g.widgets.setRight(50);

		r.noteOverlayClearRestrictedRanges();


		b.each(function(line,y){

			var widgets = [], siblings, block, rect, t, count=0;

			line.controls = line.controls || cT.append(g.controls,[],true);
			line.widgets = line.widgets || wT.append(g.widgets,[],true);

			(new Ext.CompositeElement([line.controls,line.widgets])).setTop(y+'px');

			line.each(function(o){
				r.noteOverlayAddRestrictedRange(o.getRestrictedRange(annotationOffsets));
				var c = o.getControl();
				if( c ){ c.appendTo( line.controls ); }
				if( o.hasGutterWidgets ){ widgets.push( o ); }
			});

			siblings = widgets.length-1;
			Ext.each(widgets,function(o){
				var w;
				if(count < maxAnnotations){
					count++;
					w = o.getGutterWidget(siblings);
					if(w){
						w.appendTo(line.widgets);
						w.addCls(cls);
					}
					else {
						console.warn('Annotation lied about having a gutter widget',o);
					}
				}
				else {
					sT.append(line.widgets);
					return false;//stop
				}
				return true; //continue
			});

			if(!line.controls.first()){ line.controls.remove(); delete line.controls; }
			if(!line.widgets.first()){ line.widgets.remove(); delete line.widgets; }
			else{

				r.noteOverlayRegisterAddNoteNib(
					line.first().getRecord().get('applicableRange'),
					addTpl.insertFirst(
							line.widgets,[
								siblings ? 'collapsed':'expanded'
							],true), line.first().getRecord().get('ContainerId'));
			}

			block = line.widgets || line.controls || null;
			if(block){
				t = block.dom.getBoundingClientRect();
				rect = { top:t.top, bottom:t.bottom, left:t.left, right:t.right, height:t.height, width:t.width };
				rect.top = rect.top + annotationOffsets.scrollTop;
				rect.bottom = rect.bottom + annotationOffsets.scrollTop;
				r.noteOverlayAddRestrictedRange(rect);
			}

		});
	},


	suspend: function(prefix) {
		this.rendererSuspended[prefix] = true;
	},


	resume: function(prefix) {
		delete this.rendererSuspended[prefix];
	},


	render: function(prefix){
		var me = this, containers = {}, renderedCount = 0,
			cleanContent, rootContainerId,
			cloned, descs = [], cids = [], doc = null;

		if(me.rendering){
			console.warn('Render called while rendering...');
			me.events.on('finish',Ext.bind(me.render,me,arguments),me,{single:true});
			return;
		}
		if (this.rendererSuspended[prefix]) {
			return;
		}
		if(!me.gutter[prefix]){
			console.error('no gutter');
			me.events.fireEvent('rendering');
			me.events.fireEvent('finish');
			return;
		}

		if(!me.registry[prefix]){
			me.events.fireEvent('rendering');
			me.events.fireEvent('finish');
			return;//nothing to do
		}

		me.rendering = true;
		me.events.fireEvent('rendering');
		console.log('Rendering annotations '+ prefix);
		if(console.time){
			console.time('Annotation render loop '+prefix);
		}
		Ext.suspendLayouts();

		me.registry[prefix] = Ext.Array.unique(me.registry[prefix]);

		me.sorter[prefix] = me.sorter[prefix] || me.buildSorter(prefix);
		if(me.sorter[prefix]){
			Ext.Array.sort(me.registry[prefix], me.sorter[prefix]);
		}

		me.clearBuckets(prefix);

		cloned = Ext.Array.clone(me.registry[prefix]);
		Ext.each(cloned, function(o){
			var desc = o.getRecordField ? o.getRecordField('applicableRange') : null,
				cid = o.getRecordField ? o.getRecordField('ContainerId') : null;
			if(o.doc){doc = o.doc;}
			descs.push(desc);
			cids.push(cid);
		});

		if(cloned && cloned.length > 0 && doc){
			cleanContent = me.getReader(prefix).getCleanContent();
			rootContainerId = Anchors.rootContainerIdFromDocument(doc);
			Anchors.preresolveLocatorInfo(descs, doc, cleanContent, cids, rootContainerId);
			Ext.each(cloned, function(o){
				var containerNode, containerRect;
				try {
					if(!o.isVisible){
						return;
					}

					var y, b,
					c = containers[o.getContainerId()] = (containers[o.getContainerId()]||[]);

					if(o.hasGutterWidgets){
						c.push(o);
					}

					y = o.render();
					if(y === NextThought.view.annotations.Base.HIDDEN){
						return true; //continue
					}

					/*
					if(y === NextThought.view.annotations.Base.NOT_FOUND){
						//Code to stick things we can't resolve on teh container somewhere.  In general top looks better than bottom
						//even though I think we would prefer jamming them at the bottom
						console.log('Displaying annotations with unresolvable anchor information at bottom of container');
						//Consider caching these outside this loop
						containerNode = Anchors.scopedContainerNode(doc, o.getContainerId(), rootContainerId);
						containerRect = containerNode ? containerNode.getBoundingClientRect() : null;
						if(containerRect.top && containerRect.height){
							y = containerRect.top;
						}
						if(y){
							console.log('Positioned missing annotation at bottom of container', containerNode, y);
						}
					}
				    */

					b = me.getBucket(prefix,Math.ceil(y));
					if(b){
						renderedCount++;
						b.put(o);
					}
				}
				catch(e){
					console.error(o.$className,Globals.getError(e));
				}
			});
			console[renderedCount === cloned.length ? 'log' : 'warn']('Rendered '+renderedCount+'/'+cloned.length+' annotations');
			me.layoutBuckets(prefix);
		}

		me.rendering = false;
		Ext.resumeLayouts(true);
		me.events.fireEvent('finish');
		if(console.timeEnd){
			console.timeEnd('Annotation render loop '+prefix);
		}
	}

}, function(){
	var me = this,
		fn = this.prototype.render,
		timerId = {};

	me.prototype.render = function(prefix) {
		var callerScope = this;
		if (timerId[prefix]) {
			clearTimeout(timerId[prefix]);
		}
		timerId[prefix] = setTimeout(function(){ fn.call(callerScope, prefix); }, 100);
	};
});
