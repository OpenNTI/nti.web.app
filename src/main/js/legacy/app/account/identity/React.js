const Ext = require('@nti/extjs');
const { getService } = require('@nti/web-client');
const { Identity } = require('@nti/web-profiles');
const LoginActions = require('internal/legacy/login/Actions');
require('internal/legacy/overrides/ReactHarness');

const AccountActions = require('../Actions');

module.exports = exports = Ext.define(
	'NextThought.app.account.identity.React',
	{
		extend: 'NextThought.ReactHarness',
		alias: 'widget.identity-react',
		component: Identity.Session,

		setTheme(scope, knockout) {},

		initComponent: function () {
			this.callParent(arguments);
			Object.assign(this.initialConfig, {
				addHistory: true,
				baseroute: '/app',
			});
			this.AccountActions = AccountActions.create();
			this.LoginActions = LoginActions.create();
		},

		async afterRender() {
			this.callParent(arguments);
			const service = await getService();
			const supportLinks = service.getSupportLinks();

			this.mon(Ext.getBody(), {
				click: this.onBodyClicks.bind(this, supportLinks),
			});
		},

		onBodyClicks(supportLinks, e) {
			let action = e.getTarget('a[data-action]')?.dataset.action;
			const link = e.getTarget('a[href]');
			const href = link?.getAttribute('href');

			// This is used to launch support logic the same way everywhere
			// we have a link that goes to the supportContact url. (so we can
			// just emit that url in the dom and not worry about coping logic)
			if (
				supportLinks.supportContact === href &&
				supportLinks.internalSupport
			) {
				action = action || 'support';
			}

			if (action) {
				if (call(this, `do-${action}`, link, e) !== false) {
					e.stopEvent();
				}
			}
		},

		showLink(target) {
			target.setAttribute('target', '_blank');
			return false;
		},

		'do-welcomeGuide'(target) {
			this.AccountActions.showWelcomePage(target.href);
		},

		'do-about': 'showLink',

		'do-privacy'(target) {
			this.AccountActions.showPrivacy(target);
		},

		'do-privacyForMinors'(target) {
			this.AccountActions.showChildrensPrivacy(target.href);
		},

		'do-terms'(target) {
			this.AccountActions.showTermsOfService(target.href);
		},

		'do-support'(target) {
			this.AccountActions.showContactUs();
		},

		'do-helpSite': 'showLink',

		'do-impersonate'() {
			this.LoginActions.handleImpersonate();
		},

		'do-logout'() {
			this.LoginActions.handleLogout();
		},
	}
);

function call(scope, action, ...args) {
	if (typeof action === 'string') {
		return call(scope, scope[action], ...args);
	}

	return action.apply(scope, args);
}
