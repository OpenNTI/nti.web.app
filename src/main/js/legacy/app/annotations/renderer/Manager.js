var Ext = require('extjs');
var Anchors = require('../../../util/Anchors');
var Globals = require('../../../util/Globals');
var UtilLine = require('../../../util/Line');


module.exports = exports = Ext.define('NextThought.app.annotations.renderer.Manager', {
    events: new Ext.util.Observable(),
    registry: [],

    constructor: function(reader) {
		this.callParent();
		this.reader = reader;
		this.rendererSuspended = 0;
	},

    controlLineTmpl: Ext.DomHelper.createTemplate({ cls: 'controlContainer'}).compile(),
    isDebug: false,

    /**
	 * @constructor Inner class
	 */
	Bucket: function() {
		this.values = {};
		this.length = 0;

		this.free = function() {
			this.each(function(v, k, o) { delete o[k]; }, this);
			Ext.Object.each(this, function(k, o, me) { delete me[k]; });
		};

		this.put = function(o, key) {
			key = typeof key === 'number' ? key : guidGenerator();
			if (this.values[key]) {
				throw 'existing value';
			}
			this.values[key] = o;
			this.length += 1;
			return key;
		};

		this.each = function(cb, s) {
			Ext.Object.each(this.values, function(key, value, o) {
				Ext.callback(cb, s, [value, key, o]);
			});
		};

		this.get = function(key) {
			return this.values[key];
		};

		this.first = function() {
			var key;
			for (key in this.values) {
				if (this.values.hasOwnProperty(key)) {
					return this.get(key);
				}
			}
			return null;
		};
	},

    registerGutter: function(el, reader) {
		//TODO all this junk about prefixes should go away once we aren't using a singleton here...
		if (this.gutter) {
			console.warn('replacing exisiting gutter?', this.gutter);
		}
		this.gutter = el;

		//el.addCls('debug');

		el.controls = el.down('.controls');

		this.render();//renders wait for the gutter to exist
	},

    register: function(o) {
		this.registry.push(o);
		o.requestRender();
	},

    unregister: function(o) {
		var r = this.registry;
		if (r) {
			this.registry = Ext.Array.remove(r, o);
			if (this.registry.length === 0) {
				this.sorter = null;
			}
		}
	},

    getReader: function() {
		return this.reader;
	},

    buildSorter: function(prefix) {
		//Default sort will sort by lastModified
		return function(a, b) {
			var e = {get: Ext.emptyFn},
					c = 0, $a = (a.record || e).get('Last Modified'), $b = (b.record || e).get('Last Modified');
			if ($a !== $b) {
				c = $a < $b ? 1 : -1;
			}

			return c;
		};
	},

    clearBuckets: function() {
		function clear(d) {
			while (d && d.firstChild) {
				d.removeChild(d.firstChild);
			}
		}

		var g = this.gutter;

		if (this.gutter.down('.active')) {
			this.selectedLine = this.gutter.down('.active').getAttribute('data-line');
		}
		clear(g.controls.dom);

		if (this.buckets) {
			this.buckets.free();
		}
		this.buckets = new this.Bucket();
	},

    getBucket: function(line) {
		//console.debug('prefix:'+prefix, line);
		if (line < 0) {
			//bad line, don't render:
			if (this.isDebug) {
				console.error('Annotation cannot be rendered in gutter');
			}
			return null;
		}

		var c = this.buckets,
			lineTolerance = 40,
			b = c ? c.get(line) : null,
			a = Math.round(line - (lineTolerance / 2)), z = Math.round(line + (lineTolerance / 2));

		if (!b && c) {
			c.each(function(value, key) {
				var keyNum = parseInt(key, 10);
				if (keyNum >= a && keyNum <= z) {
					b = value;
					return false;
				}
				return true;
			});

			if (!b) {
				b = new this.Bucket();
				b.line = line;
				c.put(b, line);
			}
		}
		return b;
	},

    layoutBuckets: function() {
		var g = this.gutter,
			b = this.buckets,
			cT = this.controlLineTmpl;


		b.each(function(line, y) {
			var count = 0;

			line.controls = line.controls || cT.append(g.controls, [], true);
			line.controls.setStyle('top', y + 'px');
			line.controls.set({'data-line': y});

			line.each(function(o) {
				if (o.isNote) {
					++count;
				}
			});
			if (count > 0) {
				Ext.DomHelper.overwrite(line.controls, {cls: 'count', html: count});
			}

			if (!line.controls.first()) {
				line.controls.remove();
				delete line.controls;
			}
		});
	},

    suspend: function() {
		this.rendererSuspended++;
	},

    resume: function() {
		this.rendererSuspended--;
		if (this.rendererSuspended === 0 && this.renderOnResume) {
			delete this.renderOnResume;
			this.render();
		}
	},

    render: function() {
		var me = this, containers = {}, renderedCount = 0,
			cleanContent, rootContainerId,
			cloned, descs = [], cids = [], doc = null, selectedEl;

		if (me.rendering) {
			console.warn('Render called while rendering...');
			me.events.on('finish', Ext.bind(me.render, me, arguments), me, {single: true});
			return;
		}
		if (this.rendererSuspended > 0) {
			this.renderOnResume = true;
			return;
		}
		if (!me.gutter) {
			console.error('no gutter');
			me.events.fireEvent('rendering');
			me.events.fireEvent('finish', 0);
			return;
		}

		if (!me.registry) {
			me.events.fireEvent('rendering');
			me.events.fireEvent('finish', 0);
			return;//nothing to do
		}

		me.rendering = true;
		me.events.fireEvent('rendering');
		console.log('Rendering annotations');
		if (console.time) {
			console.time('Annotation render loop');
		}
		Ext.suspendLayouts();

		me.registry = Ext.Array.unique(me.registry);

		me.sorter = me.sorter || me.buildSorter();
		if (me.sorter) {
			Ext.Array.sort(me.registry, me.sorter);
		}

		me.clearBuckets();

		cloned = Ext.Array.clone(me.registry);
		Ext.each(cloned, function(o) {
			var desc = o.getRecordField ? o.getRecordField('applicableRange') : null,
					cid = o.getRecordField ? o.getRecordField('ContainerId') : null;
			if (o.doc) {
				doc = o.doc;
			}
			descs.push(desc);
			cids.push(cid);
		});

		if (cloned && cloned.length > 0 && doc) {
			cleanContent = me.getReader().getCleanContent();
			rootContainerId = Anchors.rootContainerIdFromDocument(doc);
			Anchors.preresolveLocatorInfo(descs, doc, cleanContent, cids, rootContainerId);
			Ext.each(cloned, function(o) {
				var y, b, c;
				try {
					if (!o.isVisible) {
						return;
					}

					c = containers[o.getContainerId()] = (containers[o.getContainerId()] || []);

					if (o.isNote) {
						c.push(o);
					}

					y = o.render();
					if (y !== NextThought.app.annotations.Base.HIDDEN) {
						b = me.getBucket(Math.ceil(y));
						if (b) {
							o.getRecord().set('line', b.line);
							renderedCount++;
							b.put(o);
						}
					}
				}
				catch (e) {
					console.error(o.$className, Globals.getError(e));
				}
			});
			console[renderedCount === cloned.length ? 'log' : 'warn']('Rendered ' + renderedCount + '/' + cloned.length + ' annotations');
			me.layoutBuckets();
		}

		me.rendering = false;
		Ext.resumeLayouts(true);
		me.events.fireEvent('finish', renderedCount);
		if (console.timeEnd) {
			console.timeEnd('Annotation render loop');
		}
		if (this.selectedLine) {
			selectedEl = this.gutter.down('[data-line=' + this.selectedLine + ']');
			if (selectedEl) {
				selectedEl.addCls('active');
			}
		}
	}
}, function() {
	var me = this,
			fn = this.prototype.render,
			timerId;

	me.prototype.render = function() {
		var callerScope = this;
		if (timerId) {
			clearTimeout(timerId);
		}
		timerId = setTimeout(function() {
			fn.call(callerScope);
		}, 100);
	};
});
