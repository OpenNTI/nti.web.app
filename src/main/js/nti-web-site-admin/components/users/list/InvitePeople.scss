@import '~@nti/style-common';

.site-admin-invite-people-dialog {
	background-color: white;
	width: 760px;

	.invite-error {
		padding: rem-calc(8px 30px);
		font: normal 400 0.875rem/1.25rem $body-font-family;
		color: white;
		background: var(--primary-red);

		.inline-list {
			display: inline-flex;
			align-items: center;
			flex-direction: row;
			vertical-align: bottom;
		}
	}

	.contents {
		max-height: calc(100vh - 200px);
		min-height: 411px;
		overflow-y: auto;

		.label {
			margin-top: 0.75rem;
			margin-left: 1.25rem;
			font-size: 14px;
			color: var(--tertiary-grey);
			width: 2rem;
		}

		.invite-people-to-field {
			display: flex;
			border-bottom: solid 1px #ddd;
			border-top: solid 1px #ddd;

			.token-editor {
				flex-grow: 1;
				border: none;
				min-height: 1.5rem;

				input.token {
					width: 12rem;
				}
			}

			.label {
				width: auto;
			}

			.nti-file-input {
				span.button.file-picker {
					line-height: rem-calc(30px);
					margin-top: rem-calc(5px);
				}
			}

			.file-pill-wrap {
				font-size: 0.875rem;
				padding: 0.375rem;
				background: white;
				min-height: 1.5rem;
				line-height: 1.28571;

				.file-pill {
					display: inline-block;
					background: #fbfbfb;
					border: 1px solid #e2e2e2;
					padding: 0.1875rem 0.5rem 0.3125rem 0.75rem;
					color: var(--secondary-grey);
					font-size: 0.875rem;
					line-height: 1.28571;
					margin: 0.1875rem;

					i.icon-bold-x {
						color: var(--tertiary-grey);
						cursor: pointer;
						margin-left:  0.375rem;
					}
				}
			}

		}

		.invite-people-role-field {
			display: flex;
			border-bottom: solid 1px #ddd;

			.select-box {
				.option-label {
					font-size: 14px;
				}

				.menu-label {
					line-height: 2.75rem;
					font-size: 14px;
					font-weight: 400;

					&::after {
						font-size: 0.65rem;
					}
				}

				ul {
					padding-left: 0;
					padding-right: 10px;
					font-weight: 400;

					li {
						line-height: 1.5rem;
					}
				}

				&.open {
					.menu-label.selected {
						background: white;
						color: var(--primary-grey);
					}
				}
			}
		}

		.invite-people-message-field {
			.textarea-container {
				border: none;
				box-shadow: none;
				padding: 0.75rem;
				background-color: white;
				height: 20rem;

				textarea {
					height: 100%;
				}
			}
		}
	}

	.dialog-buttons {
		button.cancel {
			text-decoration: none;
		}
	}
}
