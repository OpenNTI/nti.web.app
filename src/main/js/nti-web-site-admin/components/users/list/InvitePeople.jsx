import React from 'react';
import PropTypes from 'prop-types';
import {scoped} from '@nti/lib-locale';
import {DialogButtons, TokenEditor, SelectBox, Panels, Input, Loading} from '@nti/web-commons';
import {validate as isEmail} from 'email-validator';

const DEFAULT_TEXT = {
	people: 'People',
	title: 'Invite People'
};

const t = scoped('nti-web-site-admin.componentsusers.list.InvitePeople', DEFAULT_TEXT);


export default class InvitePeople extends React.Component {
	static propTypes = {
		store: PropTypes.object.isRequired,
		loading: PropTypes.bool

	}

	constructor (props) {
		super(props);
	}


	state = {
		role: 'learner'
	}

	onCancel = () => {
		this.props.store.hideInviteDialog();
	}

	onSave = async () => {
		const {store} = this.props;
		const {role, message, emails} = this.state;

		if(role === 'learner') {
			store.sendLearnerInvites(emails, message);
		} else {
			store.sendAdminInvites(emails, message);
		}
	}

	onToChange = (emails) => {
		this.setState({emails});
	}

	validator = (value) => {
		let errors = [];

		if(!value || !isEmail(value)) {
			errors.push('Invalid email address');
		}

		return errors;
	}

	renderToField () {
		const { emails } = this.state;

		return (
			<div className="invite-people-to-field">
				<div className="label">To</div>
				<TokenEditor
					value={emails}
					onChange={this.onToChange}
					placeholder={emails && emails.length > 0 ? 'Add more email addresses' : 'Enter an email address'}
					validator={this.validator}
					maxTokenLength={64}/>
			</div>
		);
	}

	onRoleChange = (role) => {
		this.setState({role});
	}

	renderRoleField () {
		const {role} = this.state;

		const OPTIONS = [
			{ label: 'Learner', value: 'learner'},
			{ label: 'Administrator', value: 'admin'}
		];

		return (
			<div className="invite-people-role-field">
				<div className="label">Role</div>
				<SelectBox options={OPTIONS} onChange={this.onRoleChange} value={role}/>
			</div>
		);
	}

	onMessageChange = (message) => {
		this.setState({message});
	}

	renderMessageField () {
		return (
			<div className="invite-people-message-field">
				<Input.TextArea value={this.state.message} onChange={this.onMessageChange} placeholder="Write a personal message..."/>
			</div>
		);
	}


	render () {
		const { emails } = this.state;
		const { loading } = this.props;

		const buttons = [
			{
				label: 'Cancel',
				className: 'cancel',
				onClick: !loading && this.onCancel
			},
			{
				label: 'Send',
				className: 'save',
				disabled: loading || (!emails || emails.length === 0),
				onClick: !loading && this.onSave
			}
		];


		return (
			<div className="site-admin-invite-people-dialog">
				<div className="title">
					<Panels.TitleBar title={t('title')} iconAction={!loading && this.onCancel} />
				</div>
				<div className="contents">
					{loading && <Loading.Mask/>}
					{!loading && this.renderToField()}
					{!loading && this.renderRoleField()}
					{!loading && this.renderMessageField()}
				</div>
				<DialogButtons buttons={buttons}/>
			</div>
		);
	}
}
