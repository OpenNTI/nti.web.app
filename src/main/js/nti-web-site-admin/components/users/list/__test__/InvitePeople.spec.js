/* eslint-env jest */
import React from 'react';
import renderer from 'react-test-renderer';

import InvitePeople from '../InvitePeople';

describe('Site admin user table invite dialog test', () => {
	test('General Snapshot', () => {
		const colCmp = renderer.create(<InvitePeople loading={false} hideInviteDialog={() => {}} sendLearnerInvites={() => {}} sendAdminInvites={() => {}} />);

		const tree = colCmp.toJSON();

		expect(tree).toMatchSnapshot();
	});
});
