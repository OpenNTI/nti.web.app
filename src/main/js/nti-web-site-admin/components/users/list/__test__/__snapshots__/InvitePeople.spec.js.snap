// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Site admin user table invite dialog test General Snapshot 1`] = `
<div
  className="site-admin-invite-people-dialog"
>
  <div
    className="title"
  >
    <div
      className="panel-title-bar"
      panel-title-bar="true"
    >
      <h1>
        Invite People
      </h1>
      <i
        className="icon-light-x"
        data-tip="Close"
        onClick={[Function]}
      />
    </div>
  </div>
  <div
    className="contents"
  >
    <div
      className="invite-people-to-field"
    >
      <div
        className="label"
      >
        To
      </div>
      <div
        className="token-editor"
        onClick={[Function]}
      >
        <span
          className="placeholder"
        >
          Enter an email address
        </span>
        <input
          className="token"
          maxLength={64}
          onBlur={[Function]}
          onChange={[Function]}
          onKeyDown={[Function]}
          value=""
        />
      </div>
      <div
        className="nti-file-input"
      >
        <span
          accept=".csv"
          aria-label="Upload CSV File"
          className="button file-picker"
          data-tip="Upload CSV File"
          onChange={[Function]}
          onDrop={[Function]}
          role="button"
        >
          <input
            accept=".csv"
            onChange={[Function]}
            onDrop={[Function]}
            type="file"
          />
          <span>
            <i
              className="small icon-upload"
            />
            <span>
              Upload CSV File
            </span>
          </span>
        </span>
        <span
          className="file-select-filename"
        >
          <span
            className="nofile"
          >
            
          </span>
        </span>
      </div>
    </div>
    <div
      className="invite-people-role-field"
    >
      <div
        className="label"
      >
        Role
      </div>
      <div
        className="nti-select-native-wrapper"
      >
        <span
          className="icon-chevron-down small"
          role="presentation"
        />
        <span
          className="nti-select-value"
          role="presentation"
        >
          Learner
        </span>
        <select
          className="nti-select-native invite-select"
          onChange={[Function]}
          value="learner"
        >
          <option
            value="learner"
          >
            Learner
          </option>
          <option
            value="admin"
          >
            Administrator
          </option>
        </select>
      </div>
    </div>
    <div
      className="invite-people-message-field"
    >
      <div
        className="textarea-container"
      >
        <textarea
          className="textarea textarea-text"
          onChange={[Function]}
          onKeyDown={[Function]}
          placeholder="Write a personal message..."
          type="text"
          value=""
        />
      </div>
    </div>
  </div>
  <div
    className="dialog-buttons"
  >
    <button
      className="nti-button button save disabled primary plain"
      onClick={[Function]}
      onKeyDown={[Function]}
      role="button"
      tabIndex="0"
    >
      Send
    </button>
    <button
      className="nti-button button cancel secondary plain"
      onClick={[Function]}
      onKeyDown={[Function]}
      role="button"
      tabIndex="0"
    >
      Cancel
    </button>
  </div>
</div>
`;
